libdino/src/dbus/status_notifier_item.vala

```diff
@@ -0,0 +1,388 @@
+    [DBus (name = "StatusNotifierItemUnregistered")]
+    public signal void item_unregistered(string service);
+    [DBus (name = "StatusNotifierHostRegistered")]
+    public signal void host_registered();
+    [DBus (name = "StatusNotifierHostUnregistered")]
+    public signal void host_unregistered();
+}
+
+[DBus (name = "org.kde.StatusNotifierItem")]
+private class DBusStatusNotifierItem : Object {
+    private StatusNotifierItem item;
+    private ObjectPath menu_path;
+
+    public DBusStatusNotifierItem(StatusNotifierItem item, ObjectPath? menu_path) {
+        this.item = item;
+        this.menu_path = menu_path ?? new ObjectPath("/NO_DBUSMENU");
+    }
+
+    [DBus (name = "Category")]
+    public string category { owned get { return item.category; } }
+    [DBus (name = "Id")]
+    public string id { owned get { return item.id; } }
+    [DBus (name = "Title")]
+    public string title { owned get { return item.title; } }
+    [DBus (name = "Status")]
+    public string status { owned get { return item.status; } }
+    [DBus (name = "WindowId")]
+    public int32 window_id { get { return item.window_id; } }
+    [DBus (name = "IconName")]
+    public string icon_name { owned get { return item.icon_name; } }
+    [DBus (name = "OverlayIconName")]
+    public string overlay_icon_name { owned get { return item.overlay_icon_name ?? ""; } }
+    [DBus (name = "AttentionIconName")]
+    public string attention_icon_name { owned get { return item.attention_icon_name ?? icon_name; } }
+    [DBus (name = "ToolTip")]
+    public DBusStatusNotifierItemToolTip tool_tip { owned get { return item.tool_tip ?? DBusStatusNotifierItemToolTip.empty(); } }
+    [DBus (name = "ItemIsMenu")]
+    public bool is_menu { get { return item.is_menu; } }
+    [DBus (name = "Menu")]
+    public ObjectPath menu { get { return menu_path; } }
+
+    [DBus (name = "ContextMenu")]
+    public void context_menu(int x, int y) { item.context_menu(x, y); }
+    [DBus (name = "Activate")]
+    public void activate(int x, int y) { item.activate(x, y); }
+    [DBus (name = "SecondaryActivate")]
+    public void secondary_activate(int x, int y) { item.secondary_activate(x, y); }
+    [DBus (name = "Scroll")]
+    public void scroll(int delta, string orientation) { item.scroll(delta, orientation); }
+
+    [DBus (name = "NewTitle")]
+    public signal void on_new_title();
+    [DBus (name = "NewIcon")]
+    public signal void on_new_icon();
+    [DBus (name = "NewMenu")]
+    public signal void on_new_menu();
+    [DBus (name = "NewToolTip")]
+    public signal void on_new_tool_tip();
+    [DBus (name = "NewStatus")]
+    public signal void on_new_status(string status);
+}
+
+[DBus (name = "com.canonical.dbusmenu")]
+private class DBusMenu : Object {
+    private StatusNotifierItem item;
+
+    public DBusMenu(StatusNotifierItem item) {
+        this.item = item;
+    }
+
+    [DBus (name = "TextDirection")]
+    public string text_direction { get {
```

Comment:
I think it would probably be better to have `text_direction` exposed via the `StatusNotifierItem` class (like most properties in `DBusStatusNotifierItem`) and then also set it in `setup_systray()`. In the `setup_systray()` function you have access to the [`Gtk.get_locale_direction()`](https://valadoc.org/gtk+-3.0/Gtk.get_locale_direction.html).

Beside that, [`Pango.find_base_dir()`](https://valadoc.org/pango/Pango.find_base_dir.html) probably does a better job in determining text direction from a string :wink:

Reply:
k, il get to it